File: cp/C_Vasya_and_Robot.py
def make_move(pos, move, dis) -> None:
    if move == 'D':
        pos[1] -= dis
    elif move == 'U':
        pos[1] += dis
    elif move == 'L':
        pos[0] -= dis
    else:
        pos[0] += dis


def can_move(u, v, sln) -> bool:
    dist = abs(u[0] - v[0]) + abs(u[1] - v[1])
    if dist % 2 != sln % 2:
        return False
    return sln >= dist


def solve(s: str, n: int, x: int, y: int, sln: int) -> bool:
    pos = [0, 0]
    for i in range(sln, n):
        make_move(pos, s[i], 1)

    l, r = 0, sln
    while True:
        if can_move(pos, [x, y], sln):
            return True
        
        if r == n:
            break

        make_move(pos, s[l], 1)
        make_move(pos, s[r], -1)
        l += 1
        r += 1

    return False


if __name__ == "__main__":
    n = int(input())
    s = input()
    x, y = map(int, input().split())

    if not solve(s, n, x, y, n):
        print(-1)
    else:
        l, r = 0, n - 1
        ans = 0
        while l <= r:
            mid = (l + r) // 2
            if solve(s, n, x, y, mid):
                ans = mid
                r = mid - 1
            else:
                l = mid + 1
        print(ans)

File: cp/C_Vasya_and_Robot.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007

void move_update(pair<int, int> &pos, char move, int dir)
{
    if (move == 'U')
        pos.second += dir;
    else if (move == 'D')
        pos.second -= dir;
    else if (move == 'L')
        pos.first -= dir;
    else
        pos.first += dir;
}

bool isPossible(pair<int, int> &u, pair<int, int> v, int l)
{
    int d = abs(u.first - v.first) + abs(u.second - v.second);
    if (l % 2 != d % 2)
        return false;
    return d <= l;
}

bool solve(int len, int n, string &s, int x, int y)
{
    pair<int, int> pos = {0, 0};
    for (int i = len; i < n; i++)
    {
        move_update(pos, s[i], 1);
    }

    int l = 0, r = len;
    while (true)
    {
        if (isPossible(pos, {x, y}, len))
            return true;
        if (r == n)
            break;
        move_update(pos, s[l++], 1);
        move_update(pos, s[r++], -1);
    }
    return false;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    ll ttt = 1;
    // cin>>ttt;
    while (ttt--)
    {
        int n;
        cin >> n;
        string s;
        cin >> s;
        int x, y;
        cin >> x >> y;

        if (!solve(n, n, s, x, y))
        {
            cout << "-1\n";
            return 0;
        }

        int l = -1, r = n;
        while (r - l > 1)
        {
            int mid = (l + r) / 2;
            if (solve(mid, n, s, x, y))
                r = mid;
            else
                l = mid;
        }
        cout << r << "\n";
    }
    return 0;
}